Initial creating blueprint docs (#15816)
home-assistant/home-assistant.github.io: source/_docs/blueprint/tutorial.markdown

---

> title: "Blueprint Tutorial"

Reviewer: ```suggestion
title: "Blueprint tutorial"
```

description: "Tutorial on creating a blueprint."
---

> In this tutorial we're going to create a blueprint that controls a light based on a motion sensor. We will do this by taking an existing automation and converting it to a blueprint.
> 
> For this tutorial we use a simple automation. The process for converting a complex automation is not any different.

Reviewer: ```suggestion
In this tutorial, we're going to create a blueprint that controls a light based on a motion sensor. We will do this by taking an existing automation and converting it to a blueprint.

For this tutorial, we use a simple automation. The process for converting a complex automation is not any different.
```

{% raw %}
```yaml
trigger:
  platform: state
  entity_id: binary_sensor.motion_kitchen

action:
  service: >
    {% if trigger.to_state.state == "on" %}
      light.turn_on
    {% else %}
      light.turn_off
    {% endif %}
  target:
    entity_id: light.kitchen
```
{% endraw %}

## Create the blueprint file

Automation blueprints are YAML files (with the `.yaml` extension) and live in the `<config>/blueprints/automation/` folder. You can create as many subdirectories in this folder as you want.

To get started with our blueprint, we're going to copy the above automation YAML and save it in that directory with the name `motion_light_tutorial.yaml`.

## Add basic blueprint metadata

Home Assistant needs to know about the blueprint. This is achieved by adding a `blueprint:` section. It should contain the `domain` of the integration it is for (`automation`) and `name`, the name of your blueprint. Optionally, you can also include a `description` for your blueprint.

Add this to the top of the file:

```yaml
blueprint:
  name: Motion Light Tutorial
  domain: automation
```

## Define the configurable parts as inputs

> Now we have to decide what steps we want to make configurable. We want to make it as re-usable as possible, without losing its original intent of turning on a light based on a motion sensor.

Reviewer: ```suggestion
Now we have to decide what steps we want to make configurable. We want to make it as re-usable as possible, without losing its original intent of turning on a light-based on a motion sensor.
```

> Configurable parts in blueprints are called inputs. To make the motion sensor entity configurable, we're replacing the entity ID with a special YAML tag `!input`. This YAML tag has to be combined with the name of the input:

Reviewer: There are generally referenced to as custom tags in YAML.

```suggestion
Configurable parts in blueprints are called inputs. To make the motion sensor entity configurable, we're replacing the entity ID with a custom YAML tag `!input`. This YAML tag has to be combined with the name of the input:
```

```yaml
trigger:
  platform: state
  entity_id: !input motion_sensor
```

> For the light we can offer some more flexibility. We want to allow the user to be able to define any device or area as the target. The `target` property in the service action can contain references to areas, devices and/or entities, so that's what we will use.

Reviewer: ```suggestion
For the light, we can offer some more flexibility. We want to allow the user to be able to define any device or area as the target. The `target` property in the service action can contain references to areas, devices and/or entities, so that's what we will use.
```

Inputs are not limited to strings. They can contain complex objects too. So in this case, we're going to mark the whole `target` as input:

{% raw %}
```yaml
action:
  service: >
    {% if trigger.to_state.state == "on" %}
      light.turn_on
    {% else %}
      light.turn_off
    {% endif %}
  target: !input target_light
```
{% endraw %}

## Add the inputs to the metadata

All parts that are marked as inputs need to be added to the metadata. The minimum is that we add their names as used in the automation:

```yaml
blueprint:
  name: Motion Light Tutorial

>   domain: automation

Author: ```suggestion
  description: Turn a light on based on detected motion
  domain: automation
```

  input:
    motion_sensor:
    target_light:
```

## Use it via `configuration.yaml`

With the bare minimum metadata added, your blueprint is ready to use.

Open your `configuration.yaml` and add the following:

```yaml
automation tutorial:
  use_blueprint:
    path: motion_light_tutorial.yaml
    input:
      motion_sensor: binary_sensor.kitchen
      target_light:
        entity_id: light.kitchen
```

Reload automations and your new automation should popup. Because we configured the exact values as the original automation, they should work exactly the same.

## Adding user friendly names to the inputs

Blueprints are easier to use if it's easy to see what each field is used for. We can improve this experience by adding names and descriptions to our inputs:

```yaml
blueprint:
  name: Motion Light Tutorial
  domain: automation
  input:
    motion_sensor:
      name: Motion Sensor
      description: This sensor will be synchronized with the light.
    target_light:
      name: Lights
      description: The lights to keep in sync.
```

## Describing the inputs

> Our blueprint doesn't currently describe what the inputs should contain. Without this information Home Assistant will offer the user an empty text box.

Reviewer: ```suggestion
Our blueprint doesn't currently describe what the inputs should contain. Without this information, Home Assistant will offer the user an empty text box.
```

## The final blueprint

After we have added all the steps, our blueprint will look like this:

{% raw %}
```yaml
blueprint:
  name: Motion Light Tutorial
  domain: automation
  input:
    motion_sensor:
      name: Motion Sensor
      description: This sensor will be synchronized with the light.
      selector:
        entity:
          domain: binary_sensor
          device_class: motion
    target_light:
      name: Lights
      description: The lights to keep in sync.
      selector:
        target:
          entity:
            domain: light

trigger:
  platform: state
  entity_id: !input motion_sensor

action:
  service: >
    {% if trigger.to_state.state == "on" %}
      light.turn_on
    {% else %}
      light.turn_off
    {% endif %}
  target: !input target_light
```
{% endraw %}

## Use it via the UI

To configure it via the UI, go to **Configuration** and then **Blueprints**. Find the "Motion Light Tutorial" blueprint and click on "Create Automation".

> <p class='note'>
> Don't forget to reload automations after you make changes to your blueprint to have the UI and the automation integration pick up the latest blueprint changes.
> </p>
> 

Reviewer: ```suggestion
<div class='note'>
Don't forget to reload automations after you make changes to your blueprint to have the UI and the automation integration pick up the latest blueprint changes.
</div>

```

---
title: "Blueprint Tutorial"
description: "Tutorial on creating a blueprint."
---

> In this tutorial we're going to create a blueprint that contorls a light based on a motion sensor. We will do this by taking an existing automation and converting it to a blueprint.

Reviewer 2: ```suggestion
In this tutorial we're going to create a blueprint that controls a light based on a motion sensor. We will do this by taking an existing automation and converting it to a blueprint.
```

For this tutorial we use a simple automation. The process for converting a complex automation is not any different.

## Our automation

To create a blueprint, we first need to have a working automation with hardcoded values.

The automation for this tutorial that controls a light based off a motion sensor looks like this:

{% raw %}
```yaml
trigger:
  platform: state
  entity_id: binary_sensor.motion_kitchen

action:
  service: >
    {% if trigger.to_state.state == "on" %}
      light.turn_on
    {% else %}
      light.turn_off
    {% endif %}
  target:
    entity_id: light.kitchen
```
{% endraw %}

## Create the blueprint file

All automation blueprints live in the `<config>/blueprints/automation/` folder and are defined as YAML files (with the `.yaml` extension). You can create as many subdirectories in this folder as you want.

> To get started with our blueprint, we're giong to copy the above automation YAML and save it as the file `<config>/blueprints/automation/motion_light_tutorial.yaml`.

Reviewer 2: ```suggestion
To get started with our blueprint, we're going to copy the above automation YAML and save it as the file `<config>/blueprints/automation/motion_light_tutorial.yaml`.
```

## Add basic blueprint metadata

> Home Assistant needs to know what your blueprint is about. This is done by adding a `blueprint:` section. It should contain the `domain` of the integration it is for (`automation`) and `name`, the name of your blueprint.

Reviewer 2: ```suggestion
Home Assistant needs to know what your blueprint is about. This is done by adding a `blueprint:` section. It should contain the `domain` of the integration it is for (`automation`) and `name`, the name of your blueprint. Optionally, you can also include a `description` for your blueprint.
```

Add this to the top of the file:

```yaml
blueprint:
  name: Motion Light Tutorial

>   domain: automation

Reviewer 2: ```suggestion
  description: Turn a light on based on detected motion
  domain: automation
```

Author: I am not sure if we should include it in the example, because it takes up another line in already a long blueprint example.

Reviewer 2: For me it made sense to show that option, because I do think users should provide a description to make using blueprints easier.
 
But I also had the same thought of keeping the examples lean, which is why I only proposed to add the description twice (once in the metadata section and once at the very end), but not in the YAML examples in between. We could add a short note for that fact as well in the paragraph (à la "the subsequent YAML examples will omit the description for brevity").

Reviewer: I kinda agree with adding it. We should set the example here. When sharing blueprints, descriptions would make a difference for the consumer of such a blueprint.

Author: Yes, that makes sense. Will add.

```

## Define the configurable parts

> Now we have to decide what steps we want to make configurable. We want to make it as re-usable as possible, without losing it's original intent of turning on a light based on a motion sensor.

Reviewer 2: ```suggestion
Now we have to decide what steps we want to make configurable. We want to make it as re-usable as possible, without losing its original intent of turning on a light based on a motion sensor.
```

By limiting our blueprint to working with lights and motion sensors, we can later on tell the UI to suggest lights and motion sensors to the user, instead of all their entities.

First we're going to make the motion sensor entity configurable. We do this by replacing the entity ID with a special YAML tag `!input`. This YAML tag has to be combined with the name of the configurable part:

```yaml
trigger:
  platform: state
  entity_id: !input motion_sensor
```

For the motion sensor we just made the entity configurable. But for the light, we could offer some more flexibility. We want to allow the user to supply an area or individual lights.

To do this, we will use the `target` property in the service action. This property can contain references to areas, devices and/or entities.

Configurable parts are not limited to strings. They can contain complex objects too. So in this case, we're going to mark the whole `target` as configurable:

{% raw %}
```yaml
action:
  service: >
    {% if trigger.to_state.state == "on" %}
      light.turn_on
    {% else %}
      light.turn_off
    {% endif %}
  target: !input target_light
```
{% endraw %}

## Add the configurable parts to the metadata

> All parts that are marked as configurable need to added to the metadata. The bare minimum is that we add their names:

Reviewer 2: ```suggestion
All parts that are marked as configurable need to be added to the metadata. The bare minimum is that we add their names:
```

    path: motion_light_tutorial.yaml
    input:
      motion_sensor: binary_sensor.kitchen
      target_light:
        entity_id: light.kitchen
```

Reload automations and your new automation should popup. Because we configured the exact values as the original automation, they should work exactly the same.

## Naming the configurable parts

You can use blueprints for yourself to make reusable automations. But a great aspect of blueprints is also the possibility to share it with others.

It is possible to describe the configurable parts of the blueprint. This allows Home Assistant to automatically create a UI for it.

The first step is to add a name and description to the fields to help identify the fields.

```yaml
blueprint:
  name: Motion Light Tutorial
  domain: automation
  input:
    motion_sensor:
      name: Motion Sensor
      description: This sensor will be synchronized with the light.
    target_light:
      name: Lights
      description: The lights to keep in sync.
```

## Describing the configurable parts

Our blueprint doesn't currently describe what the configurable part should contain. Without this information Home Assistant will offer the user an empty text box.

> To do this, we will use selectors. Selectors describe a type and can be used to help the user pick a value.

Reviewer 2: ```suggestion
To instead allow Home Assistant to offer more assistance, we will use selectors. Selectors describe a type and can be used to help the user pick a matching value.
```

The selector for the motion sensor entity should describe that we want entities from the binary sensor domain that have the device class `motion`.

The selector for the target light should describe that we want to target light entities.

```yaml
blueprint:
  name: Motion Light Tutorial

>   domain: automation

Author: ```suggestion
  description: Turn a light on based on detected motion
  domain: automation
```

  input:
    motion_sensor:
      name: Motion Sensor
      description: This sensor will be synchronized with the light.
      selector:
        entity:
          domain: binary_sensor
          device_class: motion
    target_light:
      name: Lights
      description: The lights to keep in sync.
      selector:
        target:
          entity:
            domain: light
```

## The final blueprint

After we have added all the steps, our blueprint will look like this:

{% raw %}
```yaml
blueprint:
  name: Motion Light Tutorial

>   domain: automation

Reviewer 2: ```suggestion
  description: Turn a light on based on detected motion
  domain: automation
```

  input:
    motion_sensor:
      name: Motion Sensor
      description: This sensor will be synchronized with the light.
      selector:
        entity:
          domain: binary_sensor
          device_class: motion
    target_light:
      name: Lights
      description: The lights to keep in sync.
      selector:
        target:
          entity:
            domain: light

trigger:
  platform: state
  entity_id: !input motion_sensor

action:
  service: >
    {% if trigger.to_state.state == "on" %}
      light.turn_on
    {% else %}
      light.turn_off
    {% endif %}
  target: !input target_light
```
{% endraw %}

## Use it via the UI

To configure it via the UI, go to **Configuration** and then **Blueprints**. Find the "Motion Light Tutorial" blueprint and click on "Create Automation".

<p class='note'>

> Don't forget to reload automations after you make changes to your blueprint to have the UI and the automation intgration pick up the latest blueprint changes.

Reviewer 2: ```suggestion
Don't forget to reload automations after you make changes to your blueprint to have the UI and the automation integration pick up the latest blueprint changes.
```

</p>

> ![Screenshot of the blueprint  UI](/images/blueprints/tutorial-ui.png)

Reviewer 2: ```suggestion
![Screenshot of the blueprint UI](/images/blueprints/tutorial-ui.png)
```

## Share the love

The final step is to share this blueprint with others. For this tutorial we're going to share it on GitHub Gists.

- Go to [GitHub Gists](https://gist.github.com/)
- Gist description: blueprint tutorial

> - Filename including extension: `motion_light_tutorial.yaml`

Reviewer 2: ```suggestion
- File name including extension: `motion_light_tutorial.yaml`
```

Author: This is a direct copy of the text they use:

<img width="293" alt="image" src="https://user-images.githubusercontent.com/1444314/101139689-33205400-3612-11eb-87d4-458e4a0e93da.png">

Reviewer 2: Then it should stay the way the write it (although that it rather uncommon spelling according to Google). But apparently even Oxford does allow it, same with "username".

Reviewer: Yeah, that is a GitHub typo.

- <https://docs.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/f/file-name>
- <https://developers.google.com/style/word-list#letter-f>
